airflow/models/baseoperator.py

```diff
@@ -528,7 +528,7 @@ def dag(self, dag):
         elif self.has_dag() and self.dag is not dag:
             raise AirflowException(
                 "The DAG assigned to {} can not be changed.".format(self))
-        elif self.task_id not in dag.task_dict:
+        elif self.task_id:
```

Comment:
I'm missing a `if self.task_id in dag.task_dict` check and raising of `DuplicateTaskIdFound`?

Reply:
You mean like this:

```python
        elif self.task_id not in dag.task_dict:
            dag.add_task(self)
        elif self.task_id in dag.task_dict:
            raise DuplicateTaskIdFound("Task id '{}' has already been added to the DAG".format(self.task_id))
```

I removed `not in dag.task_dict` as either the task is in task_dict or not I let it go to `add_task` where we already validate it.

I am open to changing it either way so let me know which according to your makes more sense !